### PR TITLE
ddl: improve FLASHBACK DATABASE for many table case

### LINKED ISSUE #54415: Can't FLASHBACK DATABASE if there're too many tables under the database
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

drop a database with 10080 tables.

### 2. What did you expect to see? (Required)

no error

### 3. What did you see instead (Required)

```
mysql> FLASHBACK DATABASE db_99;
ERROR 8025 (HY000): entry too large, the max entry size is 6291456, the size of data is 30582644
```

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #54415

moving reading dropped TableInfo from job submit node to DDL owner, because it's too long to persist all TableInfo

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below) 100k tables under a DB failde at last step which is writting "schema diff". Will fix it in separate PR.
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/ddl.go [hunk 1]
```diff
@@ -1563,9 +1563,13 @@ type RecoverInfo struct {
 type RecoverSchemaInfo struct {
 	*model.DBInfo
 	RecoverTabsInfo []*RecoverInfo
-	DropJobID       int64
-	SnapshotTS      uint64
-	OldSchemaName   model.CIStr
+	// RecoverSchemaIDOnOwner is the new logic to avoid a large RecoverTabsInfo can't
+	// be persisted. If it's set, DDL owner will recover RecoverTabsInfo instead of
+	// job submit node.
+	RecoverSchemaIDOnOwner int64
```
Comment (D3Hunter) [inline pkg/ddl/ddl.go:1569]: seems this id duplicate with job.SchemaID and `*model.DBInfo.ID` in this struct, maybe add a `LoadOnExecute bool`?

File: pkg/executor/recover_test.go [hunk 2]
```diff
@@ -576,9 +575,15 @@ func TestFlashbackSchema(t *testing.T) {
 	tk.MustExec("insert into t_flashback values (1),(2),(3)")
 	tk.MustExec("drop database test_flashback")
 
+	// test PD connection issue causes failure after tidb_ddl_error_count_limit
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/pkg/ddl/mockClearTablePlacementAndBundlesErr", `return()`))
+	tk.MustGetErrMsg("flashback database test_flashback", "[ddl:-1]DDL job rollback, error msg: mock error for clearTablePlacementAndBundles")
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/pkg/ddl/mockClearTablePlacementAndBundlesErr", `1*return()`))
+	tk.MustExec("flashback database test_flashback")
+	require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/pkg/ddl/mockClearTablePlacementAndBundlesErr"))
```
Comment (D3Hunter) [inline pkg/executor/recover_test.go:583]:
```suggestion
	testfailpoint.Disable(t, "github.com/pingcap/tidb/pkg/ddl/mockClearTablePlacementAndBundlesErr"))
```
previous too

File: pkg/ddl/ddl_worker.go [hunk 3]
```diff
@@ -1208,7 +1208,13 @@ func (w *worker) runDDLJob(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64,
...
 	if job.IsCancelling() {
 		w.jobLogger(job).Debug("cancel DDL job", zap.String("job", job.String()))
-		return convertJob2RollbackJob(w, d, t, job)
+		ver, err = convertJob2RollbackJob(w, d, t, job)
+		if err == nil && job.State != model.JobStateRollingback && job.State != model.JobStateCancelled {
+			// must let `runDDLJob` return an error, otherwise `needUpdateRawArgs` will
+			// marshal empty Args into RawArgs.
+			err = errors.New("job can't be cancelled")
+		}
+		return ver, err
```
Comment (D3Hunter) [inline pkg/ddl/ddl_worker.go:1217]: convertJob2RollbackJob also checks job state, we can merge this part into it.

**Review by D3Hunter — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

File: pkg/ddl/table.go [hunk 4]
```diff
@@ -575,6 +568,11 @@ func (w *worker) recoverTable(t *meta.Meta, job *model.Job, recoverInfo *Recover
 	if err != nil {
 		return ver, errors.Trace(err)
 	}
+	err = clearTablePlacementAndBundles(w.ctx, recoverInfo.TableInfo)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
```
Comment (tangenta) [inline pkg/ddl/table.go:574]: This function sends a HTTP request. If there is a performance issue for `onRecoverSchema`, we can solve it by batching bundles from all involved `tableInfo`.

**Review by tangenta — APPROVED** [review]

**Pull request merged**